diff --git a/src/bamToBed/Makefile b/src/bamToBed/Makefile
index 65741ae..7e09a32 100644
--- a/src/bamToBed/Makefile
+++ b/src/bamToBed/Makefile
@@ -19,7 +19,7 @@ INCLUDES = -I$(UTILITIES_DIR)/bedFile/ \
 # ----------------------------------
 SOURCES= bamToBed.cpp
 OBJECTS= $(SOURCES:.cpp=.o)
-_EXT_OBJECTS=BamAncillary.o
+_EXT_OBJECTS=BamAncillary.o gzstream.o
 EXT_OBJECTS=$(patsubst %,$(OBJ_DIR)/%,$(_EXT_OBJECTS))
 BUILT_OBJECTS= $(patsubst %,$(OBJ_DIR)/%,$(OBJECTS))
 PROGRAM= bamToBed
